src/dict.c

```diff
@@ -1416,13 +1420,37 @@ static int _dictExpandIfNeeded(dict *d)
         (dict_can_resize != DICT_RESIZE_FORBID &&
          d->ht_used[0] / DICTHT_SIZE(d->ht_size_exp[0]) > dict_force_resize_ratio))
     {
-        if (!dictTypeExpandAllowed(d))
+        if (!dictTypeResizeAllowed(d))
             return DICT_OK;
         return dictExpand(d, d->ht_used[0] + 1);
     }
     return DICT_OK;
 }
 
+static int _dictShrinkIfNeeded(dict *d) 
+{
+    /* Incremental rehashing already in progress. Return. */
+    if (dictIsRehashing(d)) return DICT_OK;
+    
+    /* If the size of hash table is DICT_HT_INITIAL_SIZE, don't shrink it. */
+    if (DICTHT_SIZE(d->ht_size_exp[0]) == DICT_HT_INITIAL_SIZE) return DICT_OK;
+
+    /* If we reached the 1:10 ratio, and we are allowed to resize the hash
+     * table (global setting) or we should avoid it but the ratio between
+     * elements/buckets is below the lower threshold, we shrink to
+     * the next power of 2 greater than elements. */
+    if ((dict_can_resize == DICT_RESIZE_ENABLE &&  
+         d->ht_used[0] * 100 / DICTHT_SIZE(d->ht_size_exp[0]) < HASHTABLE_MIN_FILL) ||
+        (dict_can_resize != DICT_RESIZE_FORBID &&
+         d->ht_used[0] * 100 / DICTHT_SIZE(d->ht_size_exp[0]) < HASHTABLE_MIN_FILL / dict_force_resize_ratio))
+    {
+        if (!dictTypeResizeAllowed(d))
+            return DICT_OK;
+        return dictExpand(d, d->ht_used[0]);
```

Comment:
Shall we rename `dictExpand` to `dictResize` ?

Reply:
i'd rather not.. it is mentioned in too many places.

Reply:
I think it's better to keep the function names consistent with their behavior. After checking, I found that apart from `dict.c` itself, the external callers of `dictExpand` mostly intend to perform an actual expansion. How about adding another layer of encapsulation like that:

```c
/* return DICT_ERR if expand was not performed */
int dictExpand(dict *d, unsigned long size) {
    if (dictIsRehashing(d) || dictBuckets(d) >= size)
        return DICT_ERR;
    return _dictResize(d, size, NULL);
}

/* return DICT_ERR if shrink was not performed */
int dictShrink(dict *d, unsigned long size) {
    if (dictIsRehashing(d) || dictBuckets(d) <= size)
        return DICT_ERR;
    return _dictResize(d, size, NULL);
}
```

Reply:
ok, good idea

Reply:
@soloestoy I add a condition about  the number of elements in the hash table which is used in `_dictResize` before and  remove it from `_dictResize` because only `dictExpand` and `dictShrink` call `_dictResize` and both of them has this condition now. WDYT?
```C
/* return DICT_ERR if shrink was not performed */
int dictShrink(dict *d, unsigned long size) {
    /* the size is invalid if it is bigger than the size of the hash table
     * or smaller than the number of elements already inside the hash table */
    if (dictIsRehashing(d) || d->ht_used[0] > size || DICTHT_SIZE(d->ht_size_exp[0]) <= size)
        return DICT_ERR;
    return _dictResize(d, size, NULL);
}
```